examples/spells/BoundedUInt.qnt

```diff
@@ -8,258 +8,167 @@ module BoundedUInt {
+  pure def UInt(x: int): UIntT = if (isInRange(x)) Ok(x) else Err("out of range")
+
+  /// `u.bind(x => f(x))` is `f(x)` iff `u` is `Ok(x)`, i.e., if `u` is a valid bounded int wrapping `x`,
+  /// otherwise, when `u` is `Err(msg)`, it is `u`.
+  ///
+  /// ## Example
+  ///
+  /// ```
+  /// def checkedIncr(u: UintT): UintT = checkedAdd(u, UIntT(1))
+  /// run incr_UintT_twice_test = assert(UInt(0).bind(checkedIncr).bind(checkedIncr) = UInt(2))
+  /// ```
+  def bind(u: UIntT, f: (int => UIntT)): UIntT = match u {
+    | Ok(i) => f(i)
+    | Err(_) => u
     }
 
-  /// Given a bounded unsigned integer `x`, returns
-  /// true iff `x.v` lies in the `[MIN, MAX]` interval.
-  pure def isInRange(x: UIntT): bool = and {
-    x.v >= MIN,
-    x.v <= MAX
-  }
-
-  /// Given a bounded unsigned integer `x`, returns
-  /// true iff `x.v` lies in the `[MIN, MAX]` interval, and `x.error` is empty.
-  pure def isValid(x: UIntT): bool = and {
-    x.error == "",
-    isInRange(x)
-  }
-
-  /// Given a (T,T) => T operation `op`, which assumes its inputs are valid (w.r.t. `isValid`),
-  /// and two T-typed values, l and r, returns op(l,r) iff l and r are both valid. Otheriwse,
-  /// it returns some arbitrary invalid value.
-  pure def wrapErrorBin(lhs: UIntT, rhs: UIntT, op: (UIntT, UIntT) => UIntT): UIntT =
-    if (not(isValid(lhs))) lhs
-    else if (not(isValid(rhs))) rhs
-    else op(lhs, rhs)
+  /// `f.app(u, v, msg)` is `UInt(f(x,y))` if `u` is `Ok(x)` and `v` is `Ok(y)`.
+  /// If `UInt(f(x,y))` is out of range, it is `Err(msg)`.
+  pure def app(op: (int, int) => int, lhs: UIntT, rhs:UIntT, errMsg: str): UIntT =
+    lhs.bind(x => rhs.bind(y =>
+      val res = op(x, y)
+      if (isInRange(res)) Ok(res) else Err(errMsg)))
 
   /// Computes the absolute difference between lhs and rhs.
-  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
-  pure def absDiffUnsafe(lhs: UIntT, rhs: UIntT): UIntT = {
-    val res = lhs.v - rhs.v
-    { 
-      v: if (res < 0) -res else res, 
-      error: "" 
-    }
+  pure def absDiff(lhs: UIntT, rhs: UIntT): UIntT = {
+    ((x, y) => {
+      val res = x - y
+      if (res < 0) -res else res
+    }).app(lhs, rhs,
+      "impossible")
   }
 
-  /// Computes the absolute difference between lhs and rhs.
-  pure def absDiff(lhs: UIntT, rhs: UIntT): UIntT = wrapErrorBin(lhs, rhs, absDiffUnsafe)
-  
     ////////////////////////
    // CHECKED OPERATIONS //
   ////////////////////////
 
-  /// Unsafe checked integer addition. 
-  /// Computes `lhs + rhs`, setting the error field to "overflow" if overflow occurred.
-  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
-  pure def checkedAddUnsafe(lhs: UIntT, rhs: UIntT): UIntT = {
-    val res = lhs.v + rhs.v
-    { 
-      v: res,
-      error: if (res > MAX) "overflow" else ""
-    }
-  }
+  /// Checked integer addition.
+  /// Errors with "overflow"
+  pure def checkedAdd(lhs: UIntT, rhs: UIntT): UIntT = ((x,y) => iadd(x,y)).app(lhs, rhs, "overflow")
```

Comment:
Is https://github.com/informalsystems/quint/issues/1332 the reason this is not `iadd.app(...)`?

Reply:
Yes! I meant to add a comment :D Thanks!